### DISCUSSION TITLE
Performance of WSLg on nVidia Quadro M1200

### DISCUSSION THREAD

**Discussion by limingchina** [microsoft/wslg#146] [General]
Hi,

I've tried WSLg on a DELL Precision 7520, which is an nVidia Quadro M1200 as its discrete card. I've installed the driver following the README page of wslg. The version is 470.14. My windows insider build is 21370. The Ubuntu distribution I'm using is the one "Ubuntu on Windows Community Preview" mentioned here: https://ubuntu.com/blog/announcing-ubuntu-on-windows-community-preview-wsl-2. It include a Mesa 21.0.1. Using this PPA: https://launchpad.net/~kisak/+archive/ubuntu/kisak-mesa, I've upgraded to 21.0.3.

I've got the Mesa D3D12 gallium backend showing up in glxinfo like:
**Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Microsoft Corporation (0xffffffff)
    Device: D3D12 (NVIDIA Quadro M1200) (0xffffffff)
    Version: 21.0.3
    Accelerated: yes
    Video memory: 20360MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 3.3
    Max compat profile version: 3.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.0
OpenGL vendor string: Microsoft Corporation
OpenGL renderer string: D3D12 (NVIDIA Quadro M1200)
OpenGL core profile version string: 3.3 (Core Profile) Mesa 21.0.3 - kisak-mesa PPA
OpenGL core profile shading language version string: 3.30
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:**
So this means that the hardware acceleration should work. However, the performance is not what I would expect. It's pretty bad. I've tried a map application using OpenGL. The FPS is about 20FPS with 1024x768 window size. 
When I add LIBGL_ALWAYS_SOFTWARE=1 before the executable to force llvmpipe software renderer. I've got about 27FPS.

I've also tried the "Vallley" benchmark demo from https://benchmark.unigine.com/. I've only got 2-3 FPS for full screen resolution ( 1920x1080 ). The demo on native windows running normally at 40-60FPS.

Anyone has idea what could go wrong? I'd also like to mention that I've tried this PPA https://launchpad.net/~oibaf/+archive/ubuntu/graphics-drivers and get higher version of Mesa, i.e., 21.2. However, I can't get the D3D12 backend to work. It always shows the llvmpipe in _glxinfo_ output. I checked the debian package _libgl1-mesa-dri_ included in this ppa, it looks like the d3d12 gallium driver(/usr/lib/x86_64-linux-gnu/dri/d3d12_dri.so) is not included. 

Thanks,
Ming.

**Comment by spronovo** [comment]
@limingchina thanks for the report. I tried Valley and there appears to be an issue with the d3d12 backend for Mesa as you noted. I couldn't get it to work properly on Windows either. I've notified the dev working on this, this isn't expected. If you want to try things out, here's a couple of test/benchmark that works pretty well (https://www.geeks3d.com/gputest/). Are you getting good perf in application like Blender as well?

Could you point me to the map application that you used where hardware acceleration performed worst than llvmpipe?

**Comment by limingchina** [comment]
I've run some benchmarks from gputest. The result seems to be OK.

GpuTest 0.7.0
http://www.geeks3d.com
Settings:
- 1920x1080 fullscreen
- antialiasing: Off
- duration: 60000 ms

Module: FurMark
Score: 1790 points (FPS: 29)

Module: PixMark Volplosion
Score: 1259 points (FPS: 20)

Module: Plot3D
Score: 7112 points (FPS: 118)

Renderer:
- D3D12 (NVIDIA Quadro M1200)
- OpenGL: 3.1 Mesa 21.0.3 - kisak-mesa PPA


For the map application that I mentioned, it's the project from my company. Therefore I can't disclose. However, it's basically some vector map rendering.  I actually tried mapbox gl native: https://github.com/mapbox/mapbox-gl-native. It has the same problem. For this location and zoom level:  --lat="52.495300" --lon="13.356200" --zoom="10.768600" --bearing "-0.000000", with D3D12 backend, I've got around 8 FPS. When I use LIBGL_ALWAYS_SOFTWARE=1 to select the llvmpipe backend, I've got around 30FPS.
In additiona, since the "Valley" example could expose the the performance issue clearly, I guess that one could be used to debug the perf issue?

For blender, what kind of 3d model do you suggest for testing?

**Comment by limingchina** [comment]
Reported an issue for Mesa in gitlab. https://gitlab.freedesktop.org/mesa/mesa/-/issues/4761. Hope devs could pick that up and analyze the perf in more details.

**Reply by tiagoshibata** [reply]
Hi @limingchina! The merge request https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/17435 fixes the performance of Valley. Let me know if you have any more issues!

**Reply by limingchina** [reply]
I've commented on gitlab. It indeed greatly improves the performance.
